### PR TITLE
Geolocation

### PR DESCRIPTION
**Pull request by juslesan**
- Location data saved in peerInfo

- Nodes now report defined location data on status to tracker

- Leaves the option for real time updates in the future

- Tracker saves location data and has an endpoint for fetching the data

https://github.com/streamr-dev/broker/pull/172

### CODE REVIEW

File: bin/tracker.js [hunk 1]
```diff
@@ -97,7 +97,21 @@ function startServer(tracker, endpointServerPort) {
...
+        reply.send(tracker.getAllNodeLocations())
+    })
+
+    fastify.get('/location/:nodeId', async (request, reply) => {
+        if (request.params.nodeId === '') {
+            throw Error('nodeId must not be an empty string')
+        }
+        const location = tracker.getNodeLocation(request.params.nodeId)
+        if (!location) {
+            throw Error('Node not found')
```
Comment (mirpo) [inline bin/tracker.js:110]: What it does? Returns `JSON` error? Or some `HTTP` error?

File: src/composition.js [hunk 2]
```diff
@@ -9,7 +9,14 @@ const Tracker = require('./logic/Tracker')
...
-function startTracker(host, port, id = uuidv4(), maxNeighborsPerNode = 4, advertisedWsUrl = null, name) {
+const defaultLocation = {
+    latitude: null,
+    longitude: null,
+    country: null,
+    city: null
+}
+
+function startTracker(host, port, id = uuidv4(), maxNeighborsPerNode = 4, advertisedWsUrl = null, name, location = defaultLocation) {
     const peerInfo = PeerInfo.newTracker(id, name)
```
Comment (mirpo) [inline src/composition.js]: `defaultLocation` is not in use?

File: src/composition.js [hunk 3]
```diff
@@ -23,7 +30,7 @@ function startTracker(host, port, id = uuidv4(), maxNeighborsPerNode = 4, advert
     })
 }
 
-function startNetworkNode(host, port, id = uuidv4(), storages = [], advertisedWsUrl = null, name) {
+function startNetworkNode(host, port, id = uuidv4(), storages = [], advertisedWsUrl = null, name, location = defaultLocation) {
     const peerInfo = PeerInfo.newNode(id, name)
     return startEndpoint(host, port, peerInfo, advertisedWsUrl).then((endpoint) => {
```
Comment (mirpo) [inline src/composition.js:28]: Maybe location is a part of `PeerInfo`? For discussion.

File: src/logic/Tracker.js [hunk 4]
```diff
@@ -51,11 +52,13 @@ module.exports = class Tracker extends EventEmitter {
         const source = statusMessage.getSource()
         const status = statusMessage.getStatus()
         const { rtts } = status
+        const { location } = status
```
Comment (mirpo) [inline src/logic/Tracker.js]: `const { rtts, location } = status`

**Review by mirpo — CHANGES_REQUESTED** [review]
LGFM

File: src/composition.js (hunk 3 above)
Comment (juslesan) [reply src/composition.js:28]: Yes, I was thinking about this possibility also. Maybe it's a better idea to add it there

File: src/composition.js (hunk 2 above)
Comment (juslesan) [reply src/composition.js]: It's not necessary here but could be used to add locations to the trackers in the future

Comment (mirpo) [reply src/composition.js]: Let's remove it, if it's not in use now.

File: src/composition.js (hunk 3 above)
Comment (mirpo) [reply src/composition.js:28]: I think we should do it now.

File: bin/tracker.js (hunk 1 above)
Comment (juslesan) [reply bin/tracker.js:110]: Here tracker returns {"statusCode":500,"error":"Internal Server Error","message":"Node not found"} to clients

**Review by harbu — APPROVED** [review]

**Pull request merged**